forest/src/main.rs

```diff
@@ -6,23 +6,129 @@ mod daemon;
 mod logger;
 mod subcommand;
 
-use cli::{cli_error_and_die, Cli};
+use cli::{cli_error_and_die, Cli, DaemonConfig};
+
+use async_std::task;
+use daemonize_me::{Daemon, Group, User};
+use lazy_static::lazy_static;
+use log::info;
+use raw_sync::events::{Event, EventInit};
+use raw_sync::Timeout;
+use shared_memory::ShmemConf;
 use structopt::StructOpt;
+use tempfile::{Builder, TempPath};
+
+use std::fs::File;
+use std::process;
+use std::time::Duration;
+
+const EVENT_TIMEOUT: Timeout = Timeout::Val(Duration::from_secs(20));
+
+lazy_static! {
+    pub static ref IPC_PATH: TempPath = Builder::new()
+        .prefix("forest-ipc")
+        .tempfile()
+        .expect("tempfile must succeed")
+        .into_temp_path();
+}
+
+// The parent process and the daemonized child communicate through an Event in
+// shared memory. The identity of the shared memory object is written to a
+// temporary file. The parent process is responsible for cleaning up the file
+// and the shared memory object.
+fn ipc_shmem_conf() -> ShmemConf {
+    ShmemConf::new()
+        .size(Event::size_of(None))
+        .force_create_flink()
+        .flink(IPC_PATH.as_os_str())
+}
 
-#[async_std::main]
-async fn main() {
+// Initiate an Event object in shared memory.
+fn create_ipc_lock() {
+    let mut shmem = ipc_shmem_conf().create().expect("create must succeed");
+    // The shared memory object will not be deleted when 'shmem' is dropped
+    // because we're not the owner.
+    shmem.set_owner(false);
+    unsafe {
+        Event::new(shmem.as_ptr(), true).expect("new must succeed");
+    }
+}
+
+fn build_daemon<'a>(config: &DaemonConfig) -> anyhow::Result<Daemon<'a>> {
+    let mut daemon = Daemon::new()
+        .umask(config.umask)
+        .work_dir(&config.work_dir)
+        .stdout(File::create(&config.stdout)?)
+        .stderr(File::create(&config.stderr)?);
+    if let Some(user) = &config.user {
+        daemon = daemon.user(User::try_from(user)?)
+    }
+    if let Some(group) = &config.group {
+        daemon = daemon.group(Group::try_from(group)?)
+    }
+    if let Some(path) = &config.pid_file {
+        daemon = daemon.pid_file(path, Some(false))
+    }
+
+    daemon = daemon.setup_post_fork_parent_hook(|_parent_pid, _child_pid| {
+        let mut shmem = ipc_shmem_conf().open().expect("open must succeed");
+        shmem.set_owner(true);
+        let (event, _) =
+            unsafe { Event::from_existing(shmem.as_ptr()).expect("from_existing must succeed") };
+        let ret = event.wait(EVENT_TIMEOUT);
+        drop(shmem); // Delete the local link and the shared memory object.
+        if let Err(e) = ret {
+            cli_error_and_die(
+                format!(
+                    "Error unblocking process. Error was: {e}. Check the log file for details."
```

Comment:
The `e` error is quite useless. It would be better to not include it.

Reply:
Sure.